drivers/clock_control/clock_control_it8xxx2.c

```diff
@@ -0,0 +1,442 @@
+ * the very beginning.
+ *
+ * 3. I2C/SMBus Source Clock Frequency:
+ * We also strongly suggest that the I2C/SMBus Source Clock Frequency
+ * NOT be configured.
+ *
+ * So far the it8xxx2_clock_control_set_rate() only supports the MCU clock
+ * configuration but the source clock frequency of the I2C/SMBus. For there
+ * are further configurations in I2C/SMBus registers which would put the
+ * source clock frequency into division again accordingly, so it's worth
+ * noticing that the source clock frequency of the I2C/SMBus is different
+ * from the final Speed/Frequency of the I2C/SMBus.
+ *
+ * The final Sspeed/Frequency can only be obtained through related I2C/SMBus
+ * registers. Besides, the change of the I2C/SMBus Source Clock Frequency
+ * could lead that the frequency won't match the definitions in the register
+ * SCLKTS_A, that is, the speed/frequency will be wrong in such a scenario.
+ * Hence, we will still not let the configurations on I2C/SMBus Source Clock
+ * Frequency be available in options.
+ *
+ * To sum up, the it8xxx2_clock_control_set_rate() is extended to have the
+ * options in the sub_system but unfortunately it only supports the CPU
+ * clock frequency adjustment in realtime so far. However, the modifications
+ * about the sub_system are made for the possible extensions in the future.
+ *
+ */
+static int it8xxx2_clock_control_set_rate(const struct device *dev,
+					  clock_control_subsys_t sub_system,
+					  clock_control_subsys_rate_t rate)
+{
+
+	const struct clock_control_it8xxx2_config *const config = dev->config;
+	struct ecpm_it8xxx2_regs *const ecpm_regs = config->reg_base;
+	struct it8xxx2_clkctrl_subsys_t *subsys =
+		(struct it8xxx2_clkctrl_subsys_t *)(sub_system);
+
+	struct it8xxx2_clkctrl_subsys_rate_t *subsys_rate =
+		(struct it8xxx2_clkctrl_subsys_rate_t *)(rate);
+
+	uint8_t get_rate_opt = subsys->clk_opt;
+
+	uint8_t pll_clkfreq_sel_curr = ecpm_regs->PLLFREQR & PLLFREQ_MASK;
+	uint32_t pll_freq = pll_reg_to_freq[pll_clkfreq_sel_curr];
+	uint8_t mcu_current_div, divisor;
+	uint32_t s_rate = subsys_rate->clk_rate;
+	int mod;
+
+	if ((uint32_t)s_rate == 0) {
+		return -ENOTSUP;
+	}
+
+	switch (get_rate_opt) {
+	case IT8XXX2_PLL:
+		LOG_ERR("Error: Specified PLL Rate Not Supported\r\n");
+		return -ENOTSUP;
+
+	case IT8XXX2_CPU:
+		mcu_current_div = (ecpm_regs->SCDCR0 & MCUFREQ_MASK) + 1;
+		divisor = (uint8_t)(pll_freq / s_rate);
+		mod = (pll_freq % s_rate);
+
+		if ((mod != 0) || (divisor > 6) || (divisor < 1)) {
+			LOG_ERR("Error: Specified Rate Not Supported\r\n");
+			return -ENOTSUP;
+		}
+
+		if (divisor == mcu_current_div) {
+			LOG_ERR("Error: The current rate is already set.\r\n");
+			return -EALREADY;
+		}
+		ecpm_regs->SCDCR0 |= ((divisor - 1) & MCUFREQ_MASK);
+
+		break;
+
+	case IT8XXX2_SMB:
+		LOG_ERR("Error: Specified SMB Rate Not Supported\r\n");
+		return -ENOTSUP;
+
+	default:
+		LOG_ERR("Error: Specified Rate Not Supported\r\n");
+		return -ENOTSUP;
+	}
```

Comment:
If the only supported rate option is `IT8XXX2_CPU`, wouldn't it make more sense to check only against this value instead of having essentially 4 branches (`IT8XXX2_PLL`, `IT8XXX2_SMB`, `IT8XXX2_SMB` and `default`) that result in the same error code being returned?

Reply:
Well, yes. In the original revision of code, I made it only check CPU stuffs only but now for adding flexibility as @keith-zephyr mentioned, we've added several branches for the extension in the future. Simultaneously, it also extends the options in the clock_control_get_rate().

Reply:
@keith-zephyr 
Got it, sir. So I've updated it to the older one which could save some code space.